File: src/modules/payment/payment.provider.ts
// PaymentProvider is the seam between the checkout flow and a real gateway.
// The rest of the system only ever depends on this interface, so a real
// gateway (Stripe, Braintree, ...) can be dropped in by registering an
// implementation without touching order or payment services.
//
// Contract notes:
//  - Amounts are integer cents; currency is ISO 4217 (3-letter uppercase).
//  - `verifyWebhook` MUST authenticate the payload using the shared secret and
//    fail closed (throw) on any signature/format problem. The payment service
//    relies on that to reject forged callbacks.
//  - `providerReference` is the gateway's own identifier for a payment; it is
//    unique per provider and used to match webhook events back to our Payment
//    documents.

export type PaymentStatus = 'created' | 'succeeded' | 'failed' | 'refunded' | 'partially_refunded';

export interface CreatePaymentInput {
  idempotencyKey: string;
  orderId: string;
  customer: { id: string; email: string; name: string };
  amountCents: number;
  currency: string;
}

export interface CreatePaymentResult {
  providerReference: string;
  status: PaymentStatus;
  checkoutUrl: string | null;
  raw: Record<string, unknown>;
}

export interface PaymentStatusResult {
  status: PaymentStatus;
  refundedCents: number;
  raw: Record<string, unknown>;
}

export type WebhookEventType = 'payment.succeeded' | 'payment.failed' | 'refund.succeeded';

export interface WebhookEvent {
  type: WebhookEventType;
  // Gateway event id — used to deduplicate replays.
  webhookId: string;
  providerReference: string;
  amountCents: number;
  reason?: string;
  refundReference?: string;
  raw: Record<string, unknown>;
}

export interface RefundResult {
  refundReference: string;
  status: 'refunded' | 'partially_refunded';
}

export interface PaymentProvider {
  readonly name: string;
  createPayment(input: CreatePaymentInput): Promise<CreatePaymentResult>;
  getPaymentStatus(providerReference: string): Promise<PaymentStatusResult>;
  verifyWebhook(rawBody: Buffer, signature: string): Promise<WebhookEvent>;
  refundPayment(
    providerReference: string,
    input: { amountCents?: number; reason?: string },
  ): Promise<RefundResult>;
}

// Registry keyed by provider name. `getPaymentProvider` resolves the env-
// configured provider by default; callers may pass an explicit name (used by
// the webhook route, which receives the provider in the URL path).
const registry = new Map<string, PaymentProvider>();

export function registerPaymentProvider(provider: PaymentProvider): void {
  registry.set(provider.name, provider);
}

export function getPaymentProvider(name: string): PaymentProvider {
  const provider = registry.get(name);
  if (!provider) {
    throw new Error(`No payment provider registered under "${name}"`);
  }
  return provider;
}

File: java/src/main/java/com/ecommerce/backend/modules/payment/PaymentProvider.java
package com.ecommerce.backend.modules.payment;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * The seam between the checkout flow and a real gateway (mirrors
 * {@code payment.provider.ts}). Amounts are integer cents; currency is ISO 4217.
 * {@code verifyWebhook} MUST authenticate the payload and fail closed on any
 * signature/format problem, because the payment service relies on it to reject
 * forged callbacks.
 */
public interface PaymentProvider {

    String name();

    CreatePaymentResult createPayment(CreatePaymentInput input);

    PaymentStatusResult getPaymentStatus(String providerReference);

    WebhookEvent verifyWebhook(byte[] rawBody, String signature);

    RefundResult refundPayment(String providerReference, Long amountCents, String reason);

    record CreatePaymentInput(
            String idempotencyKey,
            String orderId,
            Customer customer,
            long amountCents,
            String currency) {}

    record Customer(String id, String email, String name) {}

    record CreatePaymentResult(
            String providerReference, String status, String checkoutUrl, Map<String, Object> raw) {}

    record PaymentStatusResult(String status, long refundedCents, Map<String, Object> raw) {}

    record RefundResult(String refundReference, String status) {}

    record WebhookEvent(
            String type,
            String webhookId,
            String providerReference,
            long amountCents,
            String reason,
            String refundReference,
            Map<String, Object> raw) {}

    /** Spring-managed registry keyed by provider name (mirrors the TS Map registry). */
    @org.springframework.stereotype.Component
    class Registry {

        private final Map<String, PaymentProvider> byName = new LinkedHashMap<>();

        public Registry(List<PaymentProvider> providers) {
            for (PaymentProvider provider : providers) {
                byName.put(provider.name(), provider);
            }
        }

        public void register(PaymentProvider provider) {
            byName.put(provider.name(), provider);
        }

        public PaymentProvider get(String name) {
            PaymentProvider provider = byName.get(name);
            if (provider == null) {
                throw new IllegalArgumentException(
                        "No payment provider registered under \"" + name + "\"");
            }
            return provider;
        }
    }
}
